R/dependecies.R

```diff
@@ -0,0 +1,437 @@
+  file_con <- unz(filepath, description_name)
+  dcf <- read.dcf(file = file_con)
+  close(file_con)
+  list(package = package, dcf = dcf)
+}
+
+version_from_remote <- function(package) {
+  output <- download_description_github(package)
+  unlink(package$download_path)
+  version_from_description(package
+    , output$dcf)
+}
+
+version_from_description <- function(package_name, dcf) {
+  if (!"Version" %in% colnames(dcf)) return(NA)
+  as.character(dcf[1, which(colnames(dcf) == "Version")])
+}
+
+#' Parse dependencies from description
+dependencies_from_description <- function(package, dcf) {
+  ## We install 4 kinds of dependencies listed in the description file. If our
+  ## dcf does not contain any of these elements we have no dependencies to
+  ## speak of
+  dependency_levels <- c("Depends", "Imports", "LinkingTo", "Remotes")
+  if (!any(dependency_levels %in% colnames(dcf))) return(list())
+
+  ## The parse_dcf function returns a matrix where rows correspond to packages
+  ## and column 1 corresponds to the package name and column 3 corresponds to
+  ## its version requirement.  The rownames of this matrix are the type of 
+  ## dependency (Depends, Imports, LinkingTo, or Remotes)
+  dependencies_parsed <- as.data.frame(parse_dcf(dcf
+    , depLevel = dependency_levels[dependency_levels %in%
+      colnames(dcf)])[[package$name]])
+
+  ## We separate out non-remote dependencies from remote dependencies, because
+  ## they require different logic
+  non_remote_dependencies <- dependencies_parsed[rownames(dependencies_parsed) != "Remotes", ]
+  remote_dependencies <- dependencies_parsed[rownames(dependencies_parsed) == "Remotes", ]
+
+  ## Remove the Depends entry that just corresponds to the R version 
+  ## requirements
+  non_remote_dependencies <- non_remote_dependencies[!grepl("^[rR]$"
+    , non_remote_dependencies[,1]), , drop = FALSE]
+
+  ## Parse the non-remote dependencies into a list of packages
+  if (identical(nrow(non_remote_dependencies),0L)){
+    non_remote_list <- list()
+  } else {
+    non_remote_list <- lapply(seq_along(non_remote_dependencies[,1])
+      , function(i) {
+        name <- as.character(non_remote_dependencies[i,1])
+        version <- as.character(non_remote_dependencies[i,3])
+        list(name = name, version = version)
+      })
+  }
+
+  ## Parse the remote dependencies into a list of packages
+  if (identical(nrow(remote_dependencies),0L)){
+    remote_list <- list()
+  } else {
+    matches_github <- grepl("github::"
+      , remote_dependencies[,1])
+
+    ## We do not currently support non-github remotes
+    matches_unsupported <- grepl("bitbucket::|svn::|url::|local::|gitorious"
+      , remote_dependencies[,1])
+    if (any(matches_unsupported)) {
+      stop(paste0("Package ", package$name, " from repo ", package$repo
+        , " has unsupported (non-github) remote dependencies"))
+    } else {
+      remote_list <- lapply(seq_along(remote_dependencies[,1])
```

Comment:
This is too long to put in another function! Your functions should be 10 lines of code max.